sphinx/testing/_internal/util.py

```diff
@@ -0,0 +1,115 @@
+# fmt: on
+def make_unique_id(prefix: str | os.PathLike[str] | None = None) -> str:  # NoQA: E302
+    r"""Generate a random unique identifier prefixed by *prefix*.
+
+    :param prefix: An optional prefix to prepend to the unique identifier.
+    :return: A random unique identifier.
+    """
+    suffix = os.urandom(UID_BUFLEN).hex()
+    return '-'.join((os.fsdecode(prefix), suffix)) if prefix else suffix
+
+
+def get_objects_checksum(*args: Any, **kwargs: Any) -> int:
+    """Compute a CRC-32 checksum of arbitrary objects.
+
+    The order of the positional arguments and keyword arguments matters
+    when computing the checksum, hence it is recommended to only use
+    keyword arguments whenever possible.
+
+    If an object cannot be pickled, its representation is based on the value
+    of its :func:`id`, possibly making the checksum distinct for equivalent
+    but non-pickable objects.
+    """
+    def default_encoder(x: object) -> str:
+        try:
+            return pickle.dumps(x, protocol=pickle.HIGHEST_PROTOCOL).hex()
+        except (NotImplementedError, TypeError, ValueError):
+            return format(id(x), 'x')
+
+    # use the most compact JSON format
+    env = json.dumps([args, kwargs], separators=(',', ':'),
+                     default=default_encoder, sort_keys=True)
+    # avoid using unique_object_id() since we do not really need SHA-1 entropy
+    return binascii.crc32(env.encode('utf-8'))
+
+
+@lru_cache(maxsize=65536)
+def unique_object_id(name: str) -> str:
+    """Get a unique hexadecimal identifier for an object name.
+
+    :param name: The name of the object to get a unique ID of.
+    :return: A unique hexadecimal identifier for *name*.
+    """
+    from hashlib import sha1
+
+    # ensure that non UTF-8 characters are supported and handled similarly
+    h = sha1(name.encode('utf-8', errors='backslashreplace'))
+    byt = int.from_bytes(h.digest()[:UID_BUFLEN], byteorder='little')
+    return f'%00{UID_HEXLEN}x' % byt
+
+
+def get_container_id(node: PytestNode) -> str:
+    """Get a unique identifier for the node's container.
+
+    The node's container is defined by all but the last component of the
+    node's path (e.g., ``pkg.mod.test_func`` is contained in ``pkg.mod``).
+    """
+    def get_obj_name(subject: PytestNode) -> str | None:
+        if isinstance(subject, pytest.Package):
+            return subject.name
+        if isinstance(subject, (pytest.Module, pytest.Class)):
+            return getattr(subject.obj, '__name__', None)
+        return None
+
+    names = map(get_obj_name, node.listchain())
+    container = '@'.join(filter(None, names)) or node.nodeid
+    return unique_object_id(container)
+
+
+def get_location_id(location: TestNodeLocation) -> str:
+    """Get a unique hexadecimal identifier out of a test location.
+
+    The ID is based on the physical node location (file and line number).
+    """
+    fspath, lineno = location
```

Comment:
Perhaps it doesn't matter in this case, but I noticed that sometimes we add `1` to the line number, and sometimes not.

https://github.com/picnixz/sphinx/blob/f7d7b738602d2ba8675880172587a3ff6f7bf78f/sphinx/testing/_internal/pytest_util.py#L363

Is there a potential for off-by-one in some cases?  For an ID generation scheme, I guess it probably usually doesn't matter so long as the entire set of output values retains distinct values regardless of offset.

Reply:
It doesn't really matter. The +1 is when you need a visual confirmation (since the location is 0-based) and is really when you know that the line number is correct (i.e., >= 0). Here, this function can *technically* be called with a negative number, like '-1' in some rare cases which is why I did not include the offset.

The problem is that pytest does not tell me when I'm guaranteed to have a correct test location... so I can't really know about this. Now, I should perhaps change the type and not use a test node location but just a `tuple[str, int]` since the line number could be '-1' if I'm using directly `item.location[:2]`.

Ok, long story short: 

- you can call this function with `(fspath, lineno)` when the location was successfully retrieved using `get_node_location` (because if lineno is bad, that function returns None).
- you can also call this function with `(fspath, -1)` when you are dealing with a `pytest.Item` for which `item.location[:2]` *may* return `(fspath, None)`. So, what I do when I'm in this case is replace the None by -1.

Note: in #12106, I did not correctly replace None by -1 so I'll fix this in a moment.